sw/device/lib/dif/dif_spi_device.h

```diff
@@ -0,0 +1,233 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_DIF_DIF_SPI_DEVICE_H_
+#define OPENTITAN_SW_DEVICE_LIB_DIF_DIF_SPI_DEVICE_H_
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/mmio.h"
+
+/**
+ * Represents a signal edge type.
+ */
+typedef enum dif_spi_device_edge {
+  kDifSpiDeviceEdgePositive,
+  kDifSpiDeviceEdgeNegative,
+} dif_spi_device_edge_t;
+
+/**
+ * Represents a signal edge type.
+ */
+typedef enum dif_spi_device_bit_order {
+  kDifSpiDeviceBitOrderMsbToLsb,
+  kDifSpiDeviceBitOrderLsbToMsb,
+} dif_spi_device_bit_order_t;
+
+typedef enum dif_spi_device_irq_state {
+  kDifSpiDeviceIrqStateEnable = true,
+  kDifSpiDeviceIrqStateDisable = false,
+} dif_spi_device_irq_state_t;
+
+/**
+ * Represents the types of interrupts that the SPI device will
+ * fire to signal state.
+ */
+typedef enum dif_spi_device_irq_type {
+  kDifSpiDeviceIrqTypeRxFull,
+  kDifSpiDeviceIrqTypeRxAboveLevel,
+  kDifSpiDeviceIrqTypeTxBelowLevel,
+  kDifSpiDeviceIrqTypeRxError,
+  kDifSpiDeviceIrqTypeRxOverflow,
+  kDifSpiDeviceIrqTypeTxUnderflow,
+} dif_spi_device_irq_type_t;
+
+/**
+ * Errors returned by SPI DIF functions.
+ */
+typedef enum dif_spi_device_result {
+  /**
+   * Indicates a successful operation.
+   */
+  kDifSpiDeviceResultOk = 0,
+  /**
+   * Indicates a failed precondition on the function arguments.
+   */
+  kDifSpiDeviceResultBadArg,
+} dif_spi_device_result_t;
+
+/**
+ * The length of the SPI device SRAM, in bytes.
+ *
+ * Useful for initializing FIFO lengths.
+ */
+extern const uint16_t kDifSpiDeviceSramLen;
+
+/**
+ * Configuration for initializing a SPI device.
+ */
+typedef struct dif_spi_device_config {
+  mmio_region_t base_addr;
+  dif_spi_device_edge_t clock_polarity;
+  dif_spi_device_edge_t data_phase;
+  dif_spi_device_bit_order_t tx_order;
+  dif_spi_device_bit_order_t rx_order;
+  uint16_t rx_fifo_timeout;  // uint12_t
```

Comment:
This field needs some more description.

Reply:
Shouldn't need one now, since it's just a u8.